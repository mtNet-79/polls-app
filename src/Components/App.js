import { handleInitialData } from "../actions/shared";
import PollPage from "./PollPage";
import { useState, useEffect, Fragment } from "react";
import Dashboard from "./Dashboard";
import { connect } from "react-redux";
import LogIn from "./LogIn";
import Leaderboard from "./Leaderboard";
import AddUser from "./AddUser";
import LoadingBar from "react-redux-loading-bar";
import { Routes, Route } from "react-router-dom";
import CreatePoll from "./CreatePoll";
// import { useLocation } from "react-router-dom";
import NotFound from "./NotFound";
import RequireAuth from "./wrappers/RequireAuth";
import Layout from "./wrappers/Layout";
import PropTypes from 'prop-types';

const App = (props) => {
  // console.log("app props: ", props);
  const [image, setImage] = useState("");
  const { dispatch, authedUser, loading } = props;
  useEffect(() => {
    // document.body.style.backgroundImage =  "none";
    const generateRandomAvatar = () => {
      fetch("https://avatar-endpoint.herokuapp.com/api/").then((res) => {
        res.blob().then((blobRes) => {
          const data = blobRes;
          const urlCreator = window.URL || window.webkitURL;
          const urlObj = urlCreator.createObjectURL(data);
          setImage(urlObj);
          dispatch(handleInitialData());
        });
      });
    };

    const alreadyLoggedIn = () => {
      dispatch(handleInitialData());
    };

    authedUser === null ? generateRandomAvatar() : alreadyLoggedIn();
  }, [dispatch, authedUser]);

  return (
    <Fragment>
      <LoadingBar />

      {loading ? null : (
        <div className="container">
          <Fragment>
            <Routes>
              <Route path="/add-user" element={<AddUser image={image} />} />
              <Route path="/login" element={<LogIn image={image} />} />
              <Route path="/" element={<Layout />}>
                <Route index element={<Dashboard />} />
                <Route element={<RequireAuth />}>
                  <Route path="add" element={<CreatePoll />} />
                  <Route path="leaderboard" element={<Leaderboard />} />
                  <Route
                    path="questions/:pid"
                    element={<PollPage image={image} />}
                  />
                </Route>
              </Route>
              <Route
                path="*"
                element={
                  <RequireAuth>
                    <NotFound />
                  </RequireAuth>
                }
              />
            </Routes>
          </Fragment>
        </div>
      )}
    </Fragment>
  );
};

App.propTypes = {
  authedUser:PropTypes.string,
  loading: PropTypes.bool.isRequired, 
}

const mapStateToProps = ({ authedUser, loading }) => {
  return {
    loading,
    authedUser
  };
};

export default connect(mapStateToProps)(App);
