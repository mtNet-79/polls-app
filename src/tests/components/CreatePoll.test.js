import { screen, waitFor, within } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import App from "../../Components/App";
import { renderWithProviders } from "./test-utils";

test('render to snapshot', async () => {

  const user = userEvent.setup();

  const { container } = renderWithProviders(<App />, {
    authedUser: "mthornton"
  });

  await waitFor(
    () =>
      expect(
        screen.getByRole("link", { name: /post/i })
      ).toBeInTheDocument(),
    {
      timeout: 2050,
    }
  );
  await user.click(screen.getByRole("link", { name: /post/i }));


  expect(screen.getByRole('heading', {name: /would you rather/i})).toBeInTheDocument();
  expect(screen.getByRole('heading', {name: /create a poll/i})).toBeInTheDocument();
  expect(screen.getByLabelText(/first option/i)).toBeInTheDocument()
  expect(screen.getByRole('textbox', {name: /first option/i})).toBeInTheDocument()
  expect(screen.getByLabelText(/second option/i)).toBeInTheDocument()
  expect(screen.getByRole('textbox', {name: /second option/i})).toBeInTheDocument()

  const submitBtn = screen.getByRole("button", { name: /submit/i });
  expect(submitBtn).toBeInTheDocument();
  expect(submitBtn).toBeDisabled();

  expect(container).toMatchSnapshot()

})

test("poll form ", async () => {
  const user = userEvent.setup();

  renderWithProviders(<App />, {
    authedUser: "mthornton"
  });

  await waitFor(
    () =>
      expect(
        screen.getByRole("link", { name: /post/i })
      ).toBeInTheDocument(),
    {
      timeout: 2050,
    }
  );
  await user.click(screen.getByRole("link", { name: /post/i }));

  expect(screen.getByRole('heading', {name: /would you rather/i})).toBeInTheDocument();
  expect(screen.getByRole('heading', {name: /create a poll/i})).toBeInTheDocument();
  expect(screen.getByLabelText(/first option/i)).toBeInTheDocument()
  expect(screen.getByRole('textbox', {name: /first option/i})).toBeInTheDocument()
  expect(screen.getByLabelText(/second option/i)).toBeInTheDocument()
  expect(screen.getByRole('textbox', {name: /second option/i})).toBeInTheDocument()

  const submitBtn = screen.getByRole("button", { name: /submit/i });
  expect(submitBtn).toBeInTheDocument();
  expect(submitBtn).toBeDisabled();

  //enter user input to first textbox
  const first_question = screen.getByRole("textbox", { name: /first option/i });
  await userEvent.type(first_question, "quest");
  expect(first_question.value).toBe("quest");
  expect(first_question.value.length).toBe(5);
  //button is still disabled
  expect(submitBtn).toBeDisabled();

  //enter password text to field - 5 characters 
  const second_question = screen.getByRole("textbox", { name: /second option/i });
  await user.type(second_question, "quest");
  expect(second_question.value.length).toBe(5);
  //disabled
  expect(submitBtn).toBeDisabled();

  //boht questions must be greater than 5 chars to enabled submit button
  await user.type(second_question, "ion two");
  expect(submitBtn).toBeDisabled();
  await user.type(first_question, "ion two");
  //enabled
  expect(submitBtn).not.toBeDisabled();

  //send user back to dashabord on submit
  await user.click(submitBtn)
   //has dashboard
   expect(screen.getByText(/new polls/i)).toBeInTheDocument()

});

// test('submitting question returns user to dashboard', async () => {
//   const user = userEvent.setup();

//   const { store } = renderWithProviders(<App />, {
//     authedUser: "mthornton"
//   });

//   await waitFor(
//     () =>
//       expect(
//         screen.getByRole("link", { name: /post/i })
//       ).toBeInTheDocument(),
//     {
//       timeout: 2050,
//     }
//   );

//   await user.click(screen.getByRole("link", { name: /post/i }));

//   expect(screen.getByRole('heading', {name: /would you rather/i})).toBeInTheDocument();
//   expect(screen.getByRole('heading', {name: /create a poll/i})).toBeInTheDocument();
//   //enter user input to first textbox
//   const first_question = screen.getByRole("textbox", { name: /first option/i });
//   await userEvent.type(first_question, "question one");

//   const second_question = screen.getByRole("textbox", { name: /second option/i });
//   await userEvent.type(second_question, "question two");

//   const submitBtn = screen.getByRole("button", { name: /submit/i });

//   await user.click(submitBtn)
//    //has dashboard
//    const dashboard = screen.getByRole("dashboard");
//    expect(dashboard).toBeInTheDocument()

// })
