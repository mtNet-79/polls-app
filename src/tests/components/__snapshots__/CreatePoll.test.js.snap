// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render to snapshot 1`] = `
<div>
  <div
    style="direction: ltr;"
  >
    <div
      class=""
      style="width: 100%; transition: width 200ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 1;"
    />
    <div
      style="display: table; clear: both;"
    />
  </div>
  <div
    class="container"
  >
    <div
      class="nav"
    >
      <ul>
        <li>
          <a
            class="nav-link"
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            class="nav-link"
            href="/leaderboard"
          >
            Leaderboard
          </a>
        </li>
        <li>
          <a
            aria-current="page"
            class="nav-link active"
            href="/add"
          >
            Post
          </a>
        </li>
      </ul>
      <ul>
        <li>
          <img
            alt="icon"
            class="avatar-small"
            src="https://avatar-endpoint.herokuapp.com/api/"
          />
          <span>
            mthornton
          </span>
        </li>
        <li>
          <a
            class="nav-link"
            href="/"
          >
            Log out
          </a>
        </li>
      </ul>
    </div>
    <h1>
      Would you rather
    </h1>
    <h3
      class="light"
    >
      Create a Poll
    </h3>
    <form
      class="add-new"
    >
      <label
        class="form-label center"
        for="optionOne"
      >
        First Option
      </label>
      <input
        id="optionOne"
        name="optionOne"
        placeholder="Option One"
        type="text"
        value=""
      />
      <label
        class="form-label center"
        for="optionTwo"
      >
        Second Option
      </label>
      <input
        id="optionTwo"
        name="optionTwo"
        placeholder="Option Two"
        type="text"
        value=""
      />
      <button
        class="btn"
        disabled=""
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
